gorush/notification_apns.go

```diff
@@ -9,6 +9,7 @@ import (
 	"path/filepath"
 	"time"
 
+	"github.com/facebookgo/limitgroup"
```

Comment:
This package is [deprecated](https://github.com/facebookarchive/limitgroup). Probably, it would be more future-proof to use `sync.WaitGroup`

Reply:
Is it deprecated? I know it's been archived, but don't see any mention of deprecation "tag". https://github.com/golang/go/wiki/Deprecated.

I suspect, its just a simple utility library that has completed development but still viable for use. LimitGroup does use a WaitGroup under the hood, it just has the extra functionality of specifying limit to the concurrency.

Theres also a swap in replacement https://github.com/remeh/sizedwaitgroup. Although IMO facebook is probably a better source even if it was archived. We could also do it ourselves, although I prefer it being a its own separate utility, and we'd probably just end up writing something at best, the same as what we're pulling down.

Reply:
Sorry, indeed it is not deprecated, though archived. Anyway, my remark was just a note of caution.

Reply:
@McPo I think we should remove this package and use WaitGroup as @yaronius said or [sizedwaitgroup](https://github.com/remeh/sizedwaitgroup) as you said. Why:
1. Package is archived
2. Last commit was in 2015
3. This package uses panics: [here](https://github.com/facebookarchive/limitgroup/blob/master/limitgroup.go#L22), [here](https://github.com/facebookarchive/limitgroup/blob/master/limitgroup.go#L32) and [here](https://github.com/facebookarchive/limitgroup/blob/master/limitgroup.go#L62)
4. This package doesn't have any version for go mod

Reply:
Why not just use the solution: https://github.com/appleboy/gorush/issues/470#issuecomment-592936029

Reply:
See the example code: https://github.com/sideshow/apns2/wiki/APNS-HTTP-2-Push-Speed#example-code I think we don't need to import any third-party package.

Reply:
@appleboy A few different reasons. I didn't want to conflate improvements. I believe the changes Ive made now still rely on a single client and already provides 1000x improvement for me. The next step would be to add a pool of clients equal to the number of cores (but I think that should be a separate). I have also made some other patches to the creation of clients to allow for multiple certs, so it was likely those changes were gonna conflict with each other, this was the minimal amount to achieve greater performance (And am currently under a time pressure so....).

The other reason is I haven't had a whole lot of exposure to golang (0 outside of this project), so if I could utilise an existing library I was less likely to get something wrong, (It also seems to me to be something the WaitGroup should probably support anyway).

My understanding of that snippet is that a channel size of 50 is created, and (lets say 3) clients are added to it. It means there are 50 different connection at any one time, shared amongst 3 clients, sending out 100000 notifications. Correct?

How about we do something like:
```
clients = make(chan *apns2.Client, PushConf.Ios.MaxConcurrentPushes)
clients <- getApnsClient(req)
```

Essentially we use the same channel blocking mechanism to achieve the result we want and get rid of the external library. It can then later utilised to support multiple clients at a time.

Would that work?

Reply:
@appleboy I have made those changes. As expected I had to fill the channel with the same client. https://github.com/McPo/gorush/commit/3e363d13a28480c4a175e5de99031091de3ce738

```
clients := make(chan *apns2.Client, PushConf.Ios.MaxConcurrentPushes)
for i := uint(0); i < PushConf.Ios.MaxConcurrentPushes; i++ {
 	clients <- getApnsClient(req)
}
```

In the example they only create 1 client per Core (as they state to keep it simple for demonstration purposes), the problem with this, is it highly underperforms still. There are two issues being conflated here:
1. The fact that you can have a client per core
2. The fact that a client have multiple connections

Im trying to solve #2 as it drastically boosts the performance. The snippet is showing how you can solve #1 (They aren't considering #2 because they expect thats whats already being done and to keep their example simple).

What this means is that we should have pool for clients equal to the number of cores. And divide the max number of allowable connections between each client, by creating a connection pool for each client.

So theres a single pool of clients, and a set of "pools"/channels for connections. We would only be using the pool of connections as a blocking mechanism to prevent too many connections being made, as opposed to retrieving and placing a connection back into the pool to be reused (We're just using it as a count reference).

That is why I suggest replacing the connection "pools" with a limitgroup/sizedwaitgroup, and then eventually go on to maintain a pool of clients for each core.

As such I strongly suggest we revert back to using SizedWaitGroup instead. Hopefully Ive explained that well enough.

Reply:
On second thought thats not going to prevent 2 requests coming in and 2x the max_concurrent connections being made. https://github.com/yaronius/gorush/blob/fb7be644b3082d73b145909593085914c6a72a7a/gorush/global.go#L28  Need to move the pool to global like @yaronius did in his fork. Still think we should tackle this in two different steps though, first allow multiple connections then spread them among multiple clients